File: src/pulsemesh/fusion.py
from __future__ import annotations

import math
import statistics
from typing import Any

from .models import FusionResult, TelemetrySeries
from .util import clamp


def median_abs_deviation(values: list[float], center: float) -> float:
    deviations = [abs(v - center) for v in values if math.isfinite(v)]
    if not deviations:
        return 1.0
    mad = statistics.median(deviations)
    return mad if mad > 1e-12 else 1.0


def gradient(values: list[float]) -> list[float]:
    n = len(values)
    if n == 0:
        return []
    if n == 1:
        return [0.0]
    out: list[float] = []
    for i in range(n):
        if i == 0:
            out.append(values[1] - values[0])
        elif i == n - 1:
            out.append(values[-1] - values[-2])
        else:
            out.append((values[i + 1] - values[i - 1]) / 2.0)
    return out


def percentile(values: list[float], pct: float) -> float:
    if not values:
        return 0.0
    vals = sorted(values)
    pos = clamp(pct, 0.0, 100.0) / 100.0 * (len(vals) - 1)
    lo = int(math.floor(pos))
    hi = int(math.ceil(pos))
    if lo == hi:
        return vals[lo]
    weight = pos - lo
    return vals[lo] * (1.0 - weight) + vals[hi] * weight


def fuse_series(series: TelemetrySeries, stability_threshold: float = 0.70) -> FusionResult:
    clean = [float(v) for v in series.values if math.isfinite(float(v))]
    if len(clean) < 4:
        raise ValueError(f"{series.profile_id} has too few finite values")

    center = statistics.median(clean)
    scale = median_abs_deviation(clean, center)
    normalized = [(v - center) / scale for v in clean]
    delta = gradient(normalized)
    coherence = [1.0 / (1.0 + abs(d)) for d in delta]

    mean = statistics.fmean(clean)
    stdev = statistics.pstdev(clean) if len(clean) > 1 else 0.0
    delta_abs = [abs(d) for d in delta]
    coherence_avg = statistics.fmean(coherence)
    volatility = statistics.fmean(delta_abs)
    spike_threshold = percentile(delta_abs, 95.0)
    spikes = [i for i, d in enumerate(delta_abs) if d >= spike_threshold and d > 0.0]
    stability_fraction = sum(1 for c in coherence if c >= stability_threshold) / len(coherence)

    drift = 0.0
    if len(normalized) >= 8:
        first = statistics.fmean(normalized[: max(2, len(normalized) // 4)])
        last = statistics.fmean(normalized[-max(2, len(normalized) // 4) :])
        drift = last - first

    anomaly_score = clamp((1.0 - coherence_avg) + min(1.0, volatility / 5.0) + min(1.0, len(spikes) / len(clean)), 0.0, 1.0)
    health_score = clamp(1.0 - anomaly_score, 0.0, 1.0)

    metrics: dict[str, Any] = {
        "points": len(clean),
        "mean": mean,
        "median": center,
        "min": min(clean),
        "max": max(clean),
        "stdev": stdev,
        "mad": scale,
        "coherence_avg": coherence_avg,
        "stability_fraction": stability_fraction,
        "volatility": volatility,
        "drift": drift,
        "spike_count": len(spikes),
        "spike_indices": spikes[:25],
        "anomaly_score": anomaly_score,
        "health_score": health_score,
        "used_live_data": series.used_live_data,
        "fallback_reason": series.fallback_reason,
    }
    return FusionResult(normalized=normalized, delta=delta, coherence=coherence, metrics=metrics)


def summarize_mesh(results: list[tuple[TelemetrySeries, FusionResult]]) -> dict[str, Any]:
    if not results:
        return {
            "sensor_count": 0,
            "live_sensor_count": 0,
            "fallback_sensor_count": 0,
            "mesh_health": 0.0,
            "mesh_coherence": 0.0,
            "highest_anomaly": None,
        }

    health = [float(r.metrics["health_score"]) for _, r in results]
    coherence = [float(r.metrics["coherence_avg"]) for _, r in results]
    live = [s for s, _ in results if s.used_live_data]
    highest = max(results, key=lambda item: float(item[1].metrics["anomaly_score"]))

    return {
        "sensor_count": len(results),
        "live_sensor_count": len(live),
        "fallback_sensor_count": len(results) - len(live),
        "mesh_health": statistics.fmean(health),
        "mesh_coherence": statistics.fmean(coherence),
        "highest_anomaly": {
            "profile_id": highest[0].profile_id,
            "label": highest[0].label,
            "provider": highest[0].provider,
            "anomaly_score": highest[1].metrics["anomaly_score"],
            "health_score": highest[1].metrics["health_score"],
        },
    }

File: agnt-plugin/fusion.js
// PulseMesh JS — Fusion Engine
// Port of Python fusion.py — robust median/MAD normalization, anomaly scoring, mesh summary

import { FusionResult, TelemetrySeries } from './models.js';
import { clamp } from './util.js';

function median(sorted) {
  const n = sorted.length;
  if (n === 0) return 0;
  if (n % 2 === 1) return sorted[Math.floor(n / 2)];
  return (sorted[n / 2 - 1] + sorted[n / 2]) / 2;
}

function medianAbsDeviation(values, center) {
  const deviations = values.map(v => Math.abs(v - center)).filter(Number.isFinite).sort((a, b) => a - b);
  if (deviations.length === 0) return 1.0;
  const mad = median(deviations);
  return mad > 1e-12 ? mad : 1.0;
}

function gradient(values) {
  const n = values.length;
  if (n === 0) return [];
  if (n === 1) return [0];
  const out = [];
  for (let i = 0; i < n; i++) {
    if (i === 0) out.push(values[1] - values[0]);
    else if (i === n - 1) out.push(values[n - 1] - values[n - 2]);
    else out.push((values[i + 1] - values[i - 1]) / 2);
  }
  return out;
}

function percentile(values, pct) {
  if (values.length === 0) return 0;
  const sorted = [...values].sort((a, b) => a - b);
  const pos = clamp(pct, 0, 100) / 100 * (sorted.length - 1);
  const lo = Math.floor(pos);
  const hi = Math.ceil(pos);
  if (lo === hi) return sorted[lo];
  const w = pos - lo;
  return sorted[lo] * (1 - w) + sorted[hi] * w;
}

export function fuseSeries(series, stabilityThreshold = 0.7) {
  const clean = series.values.map(Number).filter(Number.isFinite);
  if (clean.length < 4) throw new Error(`${series.profileId} has too few finite values (${clean.length})`);

  const sorted = [...clean].sort((a, b) => a - b);
  const center = median(sorted);
  const scale = medianAbsDeviation(clean, center);
  const normalized = clean.map(v => (v - center) / scale);
  const delta = gradient(normalized);
  const coherence = delta.map(d => 1 / (1 + Math.abs(d)));

  const mean = clean.reduce((s, v) => s + v, 0) / clean.length;
  const medianVal = median(sorted);
  const sortedCoh = [...coherence].sort((a, b) => a - b);
  const coherenceAvg = coherence.reduce((s, v) => s + v, 0) / coherence.length;
  const volatility = scale;
  const stabilityFraction = coherence.filter(c => c >= stabilityThreshold).length / coherence.length;
  const q25 = percentile(delta.map(Math.abs), 25);
  const q75 = percentile(delta.map(Math.abs), 75);
  const iqr = q75 - q25;
  const spikeThreshold = q75 + 1.5 * iqr;
  const spikeCount = delta.filter(d => Math.abs(d) > spikeThreshold).length;
  const drift = normalized[normalized.length - 1] - normalized[0];
  const anomalyScore = clamp((1 - coherenceAvg) + volatility / 5 * 0.1 + (spikeCount / clean.length), 0, 1);
  const healthScore = clamp(1 - anomalyScore, 0, 1);

  return new FusionResult({
    normalized, delta, coherence,
    mean, median: medianVal, center, scale,
    volatility, stabilityFraction, spikeCount, drift,
    anomalyScore, healthScore, stabilityThreshold,
  });
}

export function summarizeMesh(sensorStates, stabilityThreshold = 0.7) {
  const healthScores = sensorStates.map(s => s.metrics.health_score);
  const anomalyScores = sensorStates.map(s => s.metrics.anomaly_score);
  const meshHealth = healthScores.reduce((a, b) => a + b, 0) / healthScores.length;
  const worstSensor = sensorStates.reduce((w, s) => s.metrics.anomaly_score > (w?.metrics.anomaly_score || 0) ? s : w, null);

  return {
    mesh_health: meshHealth,
    highest_anomaly: worstSensor ? {
      profile_id: worstSensor.profile_id,
      label: worstSensor.label,
      anomaly_score: worstSensor.metrics.anomaly_score,
    } : null,
    sensor_count: sensorStates.length,
    fallback_sensor_count: sensorStates.filter(s => !s.used_live_data).length,
    alert_count: sensorStates.reduce((n, s) => n + (s.alerts?.length || 0), 0),
    critical_alert_count: sensorStates.reduce((n, s) => n + (s.alerts?.filter(a => a.severity === 'critical').length || 0), 0),
    stability_threshold: stabilityThreshold,
  };
}
